config/encrypt_file_test.go

```diff
@@ -0,0 +1,148 @@
+package config_test
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"io"
+
+	"code.cloudfoundry.org/cli/plugin"
+	"code.cloudfoundry.org/cli/plugin/pluginfakes"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/pivotal-cf/spring-cloud-services-cli-plugin/config"
+	"github.com/pivotal-cf/spring-cloud-services-cli-plugin/httpclient"
+	"github.com/pivotal-cf/spring-cloud-services-cli-plugin/httpclient/httpclientfakes"
+	"github.com/pivotal-cf/spring-cloud-services-cli-plugin/test_support"
+)
+
+var _ = Describe("Encrypt file", func() {
+
+	const (
+		serviceRegistryInstance = "some-service-registry"
+		plainText               = "plain-text"
+		errorText               = "to err is human"
+		accessToken             = "access-token"
+		bearerAccessToken       = "bearer " + accessToken
+		serviceURI              = "service-uri/"
+		encryptURI              = "service-uri/encrypt"
+		cipherText              = "cipher-text"
+	)
+
+	var (
+		fakeCliConnection     *pluginfakes.FakeCliConnection
+		fakeAuthClient        *httpclientfakes.FakeAuthenticatedClient
+		fakeResolver          func(cliConnection plugin.CliConnection, serviceInstanceName string, accessToken string, authClient httpclient.AuthenticatedClient) (string, error)
+		resolverAccessToken   string
+		testError             error
+		postResponse          io.ReadCloser
+		postStatusCode        int
+		postErr               error
+		output                string
+		err                   error
+		accessTokenURI        string
+		fakeResolverCallCount int
+		fakeResolverError     error
+	)
+
+	BeforeEach(func() {
+		fakeCliConnection = &pluginfakes.FakeCliConnection{}
+		fakeAuthClient = &httpclientfakes.FakeAuthenticatedClient{}
+		testError = errors.New(errorText)
+		postResponse, postStatusCode, postErr = ioutil.NopCloser(bytes.NewBufferString(cipherText)), http.StatusOK, nil
+		accessTokenURI = "access-token-uri"
+		fakeCliConnection.AccessTokenReturns(bearerAccessToken, nil)
+		fakeResolverCallCount = 0
+		fakeResolver = func(cliConnection plugin.CliConnection, serviceInstanceName string, accessToken string, authClient httpclient.AuthenticatedClient) (string, error) {
+			fakeResolverCallCount++
+			resolverAccessToken = accessToken
+			return serviceURI, fakeResolverError
+		}
+		config.DefaultResolver = fakeResolver
+		fakeResolverError = nil
+	})
+
+	JustBeforeEach(func() {
+		fakeAuthClient.DoAuthenticatedPostReturns(postResponse, postStatusCode, postErr)
+	})
+
+	It("should call the config server's /encrypt endpoint with content from a file", func() {
+		testDir := test_support.CreateTempDir()
+		testFile := test_support.CreateFile(testDir, "file-to-encrypt.txt")
+		output, err = config.Encrypt(fakeCliConnection, serviceRegistryInstance, "", testFile, fakeAuthClient)
+
+		Expect(fakeAuthClient.DoAuthenticatedPostCallCount()).Should(Equal(1))
```

Comment:
Testing the file processing as part of `Encrypt` means having to fake HTTP request handling. Did you explore the option of separating out a "middle layer" between `Encrypt` and `EncryptWithResolver` which could be tested separately while keeping the corresponding tests relatively focussed?

Reply:
I've made such an extraction with my latest commit. However...

I'd argue that there was already a middle layer in the `readFileContents()` function (albeit, contained within encrypt.go). That function is little more than a convenience wrapper around `ioutil.ReadFile()`, which converts the returned data to a string and prefixes the error text...little more than that. Any thorough testing I write around such an extraction would feel more like I'm just writing tests for `ioutil.ReadFile()`, which is obviously unnecessary.

Even if I did that, I'd feel compelled to write tests that flow through `Encrypt` for encrypting files...which is what I've already done. And, in fact, I've left the same (near-duplicated) tests in encrypt_file_test.go, because removing them would result in a loss of code coverage (at the little bit left in Encrypt that checks to make sure that there wasn't an error before attempting to encrypt the value).